clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -408,6 +428,48 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
                             !type_is_prefetch(memref.data.type),
                         "Function marker misplaced between instr and memref");
     }
+    if (memref.marker.type == TRACE_TYPE_MARKER &&
+        memref.marker.marker_type == TRACE_MARKER_TYPE_SYSCALL_TRACE_START) {
+        report_if_false(shard,
+                        TESTANY(OFFLINE_FILE_TYPE_KERNEL_SYSCALLS |
+                                    OFFLINE_FILE_TYPE_KERNEL_SYSCALL_INSTR_ONLY,
+                                shard->file_type_),
+                        "Found kernel syscall trace without corresponding file type");
+        report_if_false(shard, !shard->between_kernel_syscall_trace_markers_,
+                        "Nested kernel syscall traces are not expected");
+        report_if_false(shard, prev_was_syscall_marker_saved,
+                        "System call trace found without prior syscall marker");
+        report_if_false(shard,
+                        shard->last_syscall_marker_value_ ==
+                            static_cast<int>(memref.marker.marker_value),
+                        "Mismatching syscall num in trace start and syscall marker");
+        report_if_false(shard, shard->prev_instr_.decoding.is_syscall,
+                        "prev_instr at syscall trace start is not a syscall");
+        shard->pre_syscall_trace_instr_ = shard->prev_instr_;
+        shard->between_kernel_syscall_trace_markers_ = true;
+    }
+    if (memref.marker.type == TRACE_TYPE_MARKER &&
+        memref.marker.marker_type == TRACE_MARKER_TYPE_SYSCALL_TRACE_END) {
+        report_if_false(shard, shard->between_kernel_syscall_trace_markers_,
+                        "Found kernel syscall trace end without start");
+        report_if_false(shard,
+                        shard->last_syscall_marker_value_ ==
+                            static_cast<int>(memref.marker.marker_value),
+                        "Mismatching syscall num in trace end and syscall marker");
+        shard->between_kernel_syscall_trace_markers_ = false;
+        // For future checks, pretend that the previous instr was the instr just
+        // before the system call trace start.
```

Comment:
Hmm, but wouldn't we want to also check the transition from kernel back to user -- it should be `iret` or some branch like that.  Seems like we want to check both user-to-user-ignoring-kernel and kernel-back-to-user.

Reply:
Right. Added a TODO for now because the user-to-kernel transition is broken due to noise instructions at the end of PT traces (from the ioctl we make at post-syscall callback).